Routing Guide (#7)
bhumi1102/rails: guides/source/routing.md

The Purpose of the Rails Router
-------------------------------

Rails routing is a very powerful and flexible feature. The Rails router
dispatches incoming HTTP requests to specific controller actions in your Rails

> application based on the URL path. (It can also dispatch to a

Author: Sure..can remove the powerful/flexible part.

[Rack](rails_on_rack.html) application.) It also generates path and URL helpers
that can be used in your application (instead of hard-coding strings) based on
the resources configured in the router.

### Routing Incoming URLs to Code

When your Rails application receives an incoming request for:

```
GET /users/17
```

it asks the router to match it to a controller action. If the first matching route is:

```ruby
get '/users/:id', to: 'user#show'
```

the request is dispatched to the `users` controller's `show` action with `{ id: '17' }` in `params`.

Passing a `String` to `to:` will expect a `controller#action` format. When using a `Symbol`, the `to:` option should be replaced with `action:`. When using a `String` without a `#`, the `to:` option should be replaced with `controller:`:

```ruby
get '/users/:id', action: :show, controller: 'users'
```

NOTE: Rails uses snake_case for controller names, if you have a multiple word controller like `UserProfilesController`, you'd use `user_profiles#show` when specifying the route.

> ### Generating Paths and URLs from Code

Reviewer: This section confuses me a little bit because initially, I thought we were explaining how the automatically generated paths and URLs work in Rails routing but then in the example we use "as:" to specify a name for the route because we're using a "/get". Maybe there's some way to add clarity on that.

Author: Yes added a note to clarify that

The Router automatically generates paths and URLs that can be used throughout your application. Such as `user_path` and `user_url` in the below example.

If the route above is modified to be:

```ruby
get '/users/:id', to: 'users#show', as: 'user'
```

and your application contains this code in the controller:

```ruby
@user = User.find(params[:id])
```

and this in the corresponding view:

```erb
<%= link_to 'User Record', user_path(@user) %>
```

then the router will generate the path `/users/17` from `user_path(@user)`. Using the `user_path` helper allows you to avoid having to hard-code a path in your views.

There is also `user_url` which has a similar purpose. While `user_path` generates a relative url, `user_url` generates an absolute url (something like `https://yourdomain.com/users/17` in the above example).

### Configuring the Rails Router

> Here is an example of what routes look like in a typical Rails application, which live in `config/routes.rb`. The sections that follow will explain the different route helpers used in this file:

Reviewer: My 2c 😅 - I like the second suggestion

end
```

> Since this is a regular Ruby source file you can use all of its features to help you define your routes. (Be careful with variable names as they can clash with the DSL methods of the router.)

Reviewer: Feel free to ignore this as it's more of a personal preference. 

I preferred the initial versions here because it connects the two statements. I also find it slightly unusual (while grammatically acceptable) to have a fully independent sentence in parenthesis.

```suggestion
Since this is a regular Ruby source file you can use all of Ruby's features to help you define your routes but be careful with variable names as they can clash with the DSL methods of the router.
```

### Resources on the Web

> Browsers request pages from Rails by making a request for a URL using a specific HTTP method, such as `GET`, `POST`, `PATCH`, `PUT`, and `DELETE`. Each method is a request to perform an operation on the resource. A resource route maps related requests to methods (aka actions) in a single controller.

Reviewer: My 2c 😅 - I prefer the second.

| PATCH/PUT | /photos/:id      | photos#update     | update a specific photo                      |
| DELETE    | /photos/:id      | photos#destroy    | delete a specific photo                      |

> Because the router uses the HTTP verb *and* path to match inbound requests, four URLs map to seven different controller actions. For example, the same `photos/` path matches to `photos#index` when the verb is `GET` and `photos#create` when the verb is `POST`.

Reviewer: Starting a sentence with Because (while grammatically correct) [feels similar to But ](https://3.basecamp.com/3076981/buckets/35498807/todos/7116145100#__recording_7387347313). I'm not sure if we're avoiding it as well.

| PATCH/PUT | /photos/:id      | photos#update     | update a specific photo                      |
| DELETE    | /photos/:id      | photos#destroy    | delete a specific photo                      |

Since the router uses the HTTP verb *and* path to match inbound requests, four URLs can map to seven different controller actions. For example, the same `photos/` path matches to `photos#index` when the verb is `GET` and `photos#create` when the verb is `POST`.

> NOTE: Rails routes are matched in the order they are specified, so if you have a `resources :photos` above a `get 'photos/poll'` the `show` action's route for the `resources` line will be matched before the `get` line. Order matters in the `routes.rb` file. If you want the `photos/poll` route to match first, you'll need to move the `get` line **above** the `resources` line.

Author: Reworded this note

| PATCH/PUT | /geocoder      | geocoders#update  | update the one and only geocoder resource     |
| DELETE    | /geocoder      | geocoders#destroy | delete the geocoder resource                  |

NOTE: Singular resources map to plural controllers. For example, the `geocoder` resource maps to the `GeocodersController`.

A singular resourceful route generates these helpers:

* `new_geocoder_path` returns `/geocoder/new`
* `edit_geocoder_path` returns `/geocoder/edit`

> * `geocoder_path` returns `/geocoder`

Author: Kinda like it here, easier to read and notice and not get lost in the table.

| PATCH/PUT | /admin/articles/:id      | admin/articles#update  | admin_article_path(:id)      |
| DELETE    | /admin/articles/:id      | admin/articles#destroy | admin_article_path(:id)      |

Note that in the above example all of the paths have a `/admin` prefix by default convention for `namespace`.

#### Using Module

If instead you want to route `/articles` (without the prefix `/admin`) to `Admin::ArticlesController`, you can specify the module with a [`scope`][] block:

```ruby
scope module: 'admin' do
  resources :articles
end
```

Another way to write the above:

```ruby
resources :articles, module: 'admin'
```

#### Using Scope

Alternatively, you can also route `/admin/articles` to `ArticlesController` (without the `Admin::` module prefix). You can specify the path with a `scope` block:

```ruby
scope '/admin' do
  resources :articles
end
```

Another way to write the above:

```ruby
resources :articles, path: '/admin/articles'
```

> For both alternatives (without `/admin` in path and without `Admin::` in module prefix), the named route helpers remain the same as if you did not use `scope`.

Author: removed 'both'. Otherwise this spells out the alternatives: "(without `/admin` in path and without `Admin::` in module prefix)"

end
```

Deeply nested resources can become cumbersome to maintain and reason about. In the above example, the application would recognize paths such as:

```
/publishers/1/magazines/2/photos/3
```

> The corresponding route helper would be `publisher_magazine_photo_url`, requiring you to specify objects at all three levels. This can become confusing and overly complex. Therefore, the general rule of thumb is to only nest resources 1 level deep.

Author: Sure

#### Shallow Nesting

One way to avoid deep nesting (as recommended above) is to generate the
collection actions scoped under the parent, so as to get a sense of the

> hierarchy, but to not nest the member actions. In other words, to only build

Reviewer: Maybe it would be helpful to explicitly list out which CRUD actions (i.e. member actions) that should not be nested for beginner developers that are not quite familiar with the collection vs member terms.

Author: Added definition for "member" and "collection" action with example CRUD actions.

| PATCH/PUT | /articles/:id(.:format)                      | articles#update   | article_path             |
| DELETE    | /articles/:id(.:format)                      | articles#destroy  | article_path             |

The [`shallow`][] method with a block creates a scope inside of which every nesting is shallow. This generates the same routes as the previous example:

```ruby
shallow do
  resources :articles do
    resources :comments
    resources :quotes
  end
end
```

> There are two options that can be used with `scope` to customize shallow routes - `:shallow_path` and `:shallow_prefix`.

Reviewer: Maybe it would be useful to mention why we'd use `shallow_path` and `shallow_prefix`.

Author: Other than to customize it, as it say, no other reason I can think of off hand.

> ### Routing Concerns

Reviewer: I could not comment on the [actual line](https://github.com/bhumi1102/rails/pull/7/files#diff-42e915ace4a21890e4d38f5b6230345d62540977e922f87d04b672db908539e7R504-R508) 

```ruby
namespace :articles do
  concerns :commentable
end
```

maybe it will be useful to expand the equivalence for the namespace example as well?

Author: Added words instead, so the code blocks are not repetitive and long.

<%= link_to 'Ad details', url_for([@magazine, @ad]) %>
```

In this case, Rails will see that `@magazine` is a `Magazine` and `@ad` is an `Ad` and will therefore use the `magazine_ad_path` helper. An even shorter way to write that `link_to` is to specify just the object instead of the full `url_for` call:

> 

Reviewer: Maybe it would be useful to show the link that these examples generate?

Author: added

> <%= link_to 'Ad details', magazine_ad_path(@magazine, @ad) %>

Author: Added api link for both 'link_to' and 'url_for', as I noticed that was also missing a reference.

<%= link_to 'Magazine details', @magazine %>
```

For other actions, you need to insert the action name as the first element of the array, for `edit_magazine_ad_path`:

```erb
<%= link_to 'Edit Ad', [:edit, @magazine, @ad] %>
```

This allows you to treat instances of your models as URLs, and is a key advantage to using the resourceful style.

> NOTE: In order to automatically derive paths and URLs from objects such as `[@magazine, @ad]`, Rails uses methods from `ActiveModel::Naming` and `ActiveModel::Conversion` modules. Specifically the `@magazine.model_name.route_key` returns `magazines` and `@magazine.to_param` returns a string representation of the model's `id`. So the generated path may be something like `/magazines/1/ads/42` for the objects `[@magazine, @ad]`.

Reviewer: Let's add links for the `ActiveModel::Naming` and `ActiveModel::Conversion` modules

> NOTE: In order to automatically derive paths and URLs from objects such as `[@magazine, @ad]`, Rails uses methods from `ActiveModel::Naming` and `ActiveModel::Conversion` modules. Specifically the `@magazine.model_name.route_key` returns `magazines` and `@magazine.to_param` returns a string representation of the model's `id`. So the generated path may be something like `/magazines/1/ads/42` for the objects `[@magazine, @ad]`.

Reviewer: Also a missing comma after Specifically.

```suggestion
NOTE: In order to automatically derive paths and URLs from objects such as `[@magazine, @ad]`, Rails uses methods from `ActiveModel::Naming` and `ActiveModel::Conversion` modules. Specifically, the `@magazine.model_name.route_key` returns `magazines` and `@magazine.to_param` returns a string representation of the model's `id`. So the generated path may be something like `/magazines/1/ads/42` for the objects `[@magazine, @ad]`.
```

[ActionView::RoutingUrlFor#url_for]: https://api.rubyonrails.org/classes/ActionView/RoutingUrlFor.html#method-i-url_for

### Adding More RESTful Actions

> You are not limited to the seven routes that RESTful routing creates by default. You can add additional routes that apply to the collection or individual members of the collection.

Reviewer: Maybe a link or more details to the "seven routes" ?

> The below sections describe adding member routes and collection routes. The term `member` refers to routes acting on a single element, such as `show`, `update`, or `destroy`. The term `collection` refers to routes acting on multiple, or a collection of, elements, such as the `index` route.

Reviewer: It's great that we explain what a `member` and `collection` route is here, but we use these terms frequently in the sections above, so maybe it would be useful to introduce these definitions before the first instance where it's being used.

Author: Added definition when they are first mentioned further up too.

end
```

This will recognize URL `/photos/1/preview` with GET, and route to the `preview` action of `PhotosController`, with the resource id value in `params[:id]`. It will also create the `preview_photo_url` and `preview_photo_path` helpers.

Within the `member` block, each route definition specifies the HTTP verb (`get`

> in the above example with `get 'preview`). In addition to [`get`][], you can use

Reviewer: ```suggestion
in the above example with `get 'preview'`). In addition to [`get`][], you can use
```

### Naming Routes

You can specify a name that will used by the `_path` and `_url` helpers for any route using the `:as` option:

```ruby
get 'exit', to: 'sessions#destroy', as: :logout
```

This will create `logout_path` and `logout_url` as the route helpers in your application. Calling `logout_path` will return `/exit`.

You can also use `as` to override routing helper names defined by `resources` by placing a custom route definition *before* the resource is defined, like this:

```ruby
get ':username', to: 'users#show', as: :user
resources :users
```

This will define a `user_path` helper that will match a route such as `/jane`. Inside the `show` action of `UsersController`, `params[:username]` will contain the username for the user.

### HTTP Verb Constraints

In general, you should use the [`get`][], [`post`][], [`put`][], [`patch`][], and [`delete`][] methods to constrain a route to a particular verb. There is a [`match`][] method that you could use with the `:via` option to match multiple verbs at once:

```ruby

> match 'photos', to: 'photos#show', via: [:get, :post]

Author: added something

> ```

Reviewer 2: Should we show the results here as well?
````suggestion
# >> main_url
# => Not sure what this would return...
```
````

> Customizing Resourceful Routes

Reviewer 2: After reading about "Resourceful Routes" the customization might go unnoticed if the next section is called "Non-Resourceful Routes".

Would it make sense to move this whole section under "Resource Routing: the Rails Default"?
In this section we mention constraints though, and we currently explain constraints in "Non-Resourceful Routes".

Or should we add a last section to "Resourceful Routes" that links to this section?

Author: Sure, added a link. As the "customizing" one is a top level section, it'll be visible in the right navigation and such that should help too.

> This will route to the `Admin::UserPermissions` controller.

Reviewer 2: Should we always use the full controller name?
```suggestion
This will route to the `Admin::UserPermissionsController` instance.
```

form_with(model: [@account, @article])   # => <form action="/1/article/9" ...>
```

The `:as` option is also not required, but without it, Rails will raise an error when evaluating `url_for([@account, @article])` or other helpers that rely on `url_for`, such as [`form_with`][].

[`form_with`]: https://api.rubyonrails.org/classes/ActionView/Helpers/FormHelper.html#method-i-form_with

### Restricting the Routes Created

> By default, using `resources` creates routes for the seven default actions (`index`, `show`, `new`, `create`, `edit`, `update`, and `destroy`). You can use the `:only` and `:except` options to limit which routes are created.

Author: We do name the seven routes via the tables. I guess I can add a sentence as well to be even more explicit.

> The `:only` option tells Rails to create only the specified routes:

Author: defined 'only' further up.

Video.find_by(identifier: params[:identifier])
```

You can override `ActiveRecord::Base#to_param` of the associated model to

> construct a URL:

Reviewer: I'm not certain that it's obvious to a beginner what we're doing in this example, I can't think of a way offhand but do you think there's any more detail we can add to further explain what we're doing here and what it's helping us achieve? In addition, is there a link to the API for `ActiveRecord::Base#to_param` that we could link to?

Author: Added a bit more to clarify and link to `to_param`

edit_video_path(video) # => "/videos/Roman-Holiday/edit"
```

> Breaking Up Large a Route File With `draw`

Reviewer 2: ```suggestion
Breaking Up a Large Route File With `draw`
```

end
```

Calling `draw(:admin)` inside the `Rails.application.routes.draw` block itself
will try to load a route file that has the same name as the argument given
(`admin.rb` in this example). The file needs to be located inside the
`config/routes` directory or any sub-directory (i.e. `config/routes/admin.rb` or
`config/routes/external/admin.rb`).

NOTE: You can use the normal routing DSL inside a secondary routing file such as `admin.rb`, but *do not* surround it with the `Rails.application.routes.draw` block. That should be used in the main `config/routes.rb` file only.

[`draw`]: https://api.rubyonrails.org/classes/ActionDispatch/Routing/Mapper/Resources.html#method-i-draw

NOTE: Don't use the `draw` feature unless you really need it. Having multiple routing files make it harder to discover routes in one place. For most applications - even those with a few hundred routes - it's easier for developers to have a single routing file. The Rails routing DSL already offers a way to break routes in an organized manner with `namespace` and `scope`.

> Inspecting and Testing Routes

Reviewer 2: Should we just split this into two separate sections ("Inspecting Routes" and "Testing Routes")?
I'm not sure they are that related.

Author: sure

```

Calling `draw(:admin)` inside the `Rails.application.routes.draw` block itself
will try to load a route file that has the same name as the argument given
(`admin.rb` in this example). The file needs to be located inside the
`config/routes` directory or any sub-directory (i.e. `config/routes/admin.rb` or
`config/routes/external/admin.rb`).

NOTE: You can use the normal routing DSL inside a secondary routing file such as `admin.rb`, but *do not* surround it with the `Rails.application.routes.draw` block. That should be used in the main `config/routes.rb` file only.

[`draw`]: https://api.rubyonrails.org/classes/ActionDispatch/Routing/Mapper/Resources.html#method-i-draw

NOTE: Don't use the `draw` feature unless you really need it. Having multiple routing files make it harder to discover routes in one place. For most applications - even those with a few hundred routes - it's easier for developers to have a single routing file. The Rails routing DSL already offers a way to break routes in an organized manner with `namespace` and `scope`.

Inspecting Routes
-----------------

Rails offers a few different ways of inspecting and testing your routes.

### Listing Existing Routes

> To get a complete list of routes available in an application, visit `http://localhost:3000/rails/info/routes` in the **development** environment. You can also execute the `bin/rails routes` command in your terminal to get the same output.

Reviewer: Do you think it will be useful to mention that the `_path` and the `_url` helpers can be derived for a route by adding them as suffixes to the route name? I've seen someone mention it like that in a blog post but maybe it's too obvious - I'm not sure.

Author: Not sure what you mean...maybe you can give an example?

Author: I see! Let me see if I can include that in a concise way.

Author: Done!

Controller#Action | users#edit
```

### Routes in Rails Console

You can access route helpers using `Rails.application.routes.url_helpers` within the [Rails Console](command_line.html#bin-rails-console). For example:

```irb

> irb> helpers = Rails.application.routes.url_helpers
> => #<Module:0x00007fc1dfc97f68>
> 
> irb> helpers.users_path
> => "/users"
> 
> irb> user = User.first
> => #<User:0x00007fc1eab81628
> irb> helpers.edit_user_path(user)
> => "/users/1/edit"
> ```

Reviewer 2: Routes can be accessed using the default `app` helper:
```suggestion
irb> app.users_path
=> "/users"

irb> user = User.first
=> #<User:0x00007fc1eab81628
irb> app.edit_user_path(user)
=> "/users/1/edit"
```
```

### Searching Routes

You can search through your routes with the grep option: `-g`. This outputs any routes that partially match the URL helper method name, the HTTP verb, or the URL path.

```bash
$ bin/rails routes -g new_comment
$ bin/rails routes -g POST
$ bin/rails routes -g admin
```

> If you only want to see the routes that map to a specific controller, there's the `-c` option.

Reviewer 2: Maybe mention where the `c` stands for (although it's pretty obvious)?
```suggestion
If you only want to see the routes that map to a specific controller, there's the controller option: `-c`.
```

Controller#Action | users#edit
```

> ### Routes in Rails Console

Reviewer: Love that you added this section!

$ bin/rails routes -c Articles::CommentsController
```

TIP: The output from `bin/rails routes` is easier to read if you widen your terminal window until the output lines don't wrap.

> ### Listing Unused Routes

Reviewer: TIL!

You can scan your application for unused routes with the `--unused` option. For example:

```bash
$ rails routes --unused
Found 8 unused routes:

     Prefix Verb   URI Pattern                Controller#Action
     people GET    /people(.:format)          people#index
            POST   /people(.:format)          people#create
 new_person GET    /people/new(.:format)      people#new
edit_person GET    /people/:id/edit(.:format) people#edit
     person GET    /people/:id(.:format)      people#show
            PATCH  /people/:id(.:format)      people#update
            PUT    /people/:id(.:format)      people#update
            DELETE /people/:id(.:format)      people#destroy
```

Testing Routes
--------------

> Routes should be included in your testing strategy (just like the rest of your application). Rails offers three built-in assertions designed to make testing routes simpler:

Reviewer: Wondering if the "just like the rest of your application" in brackets is necessary?

```suggestion
Routes should be included in your testing strategy. Rails offers three built-in assertions designed to make testing routes simpler:
```

Author: Removing that whole line actually.
